File: python/prime.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

def isprime(num):
    import math
    if num < 2: return False
    if num == 2: return True
    if not num % 2:
        return False
    for i in xrange(3, int(math.sqrt(num)) + 1, 2):
        if not num % i:
            return False
    return True

def genprime():
    yield 2
    yield 3
    yield 5
    i = 1
    while True:
        res = 6 * i + 1
        if isprime(res): yield res
        res = 6 * i + 5
        if isprime(res): yield res
        i += 1

def factorize(num):
    res = []
    if isprime(num):
        res.append(num)
        return res
    for i in genprime():
        if i * 2 - 1 > num: return res
        j = num
        while True:
            if not j % i:
                res.append(i)
                j /= i
            else: break


if __name__ == '__main__':
    def timing(f):
        import time
        def new_f(*args, **kwds):
            t = time.time()
            result = f(*args, **kwds)
            print ('%s: %ss'%(f.func_name, time.time()-t))
            return result
        return new_f

    @timing
    def main():
        for i in genprime():
            if i > 1000000: break
            print i

    main()

File: scala/Prime.scala
object Prime {
  def isPrime(num:Int):Boolean = {
    if(num == 2) true
    else {
      if(num % 2 == 0) false
      else {
        for(i <- 3 until num / 2 by 2) yield {
          if(num % i == 0) {return false}
        }
        true
      }
    }
  }
  def gen(max:Int) = {
    for(i <- 2 to max if(isPrime(i))) yield i
  }
  def factorize(num:Int) = {
    def f = for(i <- gen(num / 2)) yield {if(num % i == 0){i}}
    f.filter(_ != 0)
  }
  def main(args:Array[String]) {
    println(isPrime(2))
    println(gen(1000))
    println(factorize(100))
  }
}
